ot/gromov/_utils.py

```diff
@@ -522,3 +524,254 @@ def h2(b):
+        - nx.sum(pi1 * nx.log(a)) - nx.sum(pi2 * nx.log(b))
+
+    return res
+
+
+def div(pi, pi1, pi2, a, b, divergence, nx=None):
+    """
+    Calculate Div(pi, a \otimes b).
+    """
+
+    if nx is None:
+        pi, pi1, pi2, a, b = list_to_array(pi, pi1, pi2, a, b)
+        nx = get_backend(pi, pi1, pi2, a, b)
+
+    if divergence == "kl":
+        res = approx_shortcut_kl(pi, pi1, pi2, a, b, nx) \
+            - nx.sum(pi1) + nx.sum(a) * nx.sum(b)
+    elif divergence == "l2":
+        res = (nx.sum(pi**2) + nx.sum(a**2) * nx.sum(b**2)
+               - 2 * nx.dot(a, nx.dot(pi, b))) / 2
+
+    return res
+
+
+# Support functions for KL and squared L2 between product measures:
+# Calculate Div(mu \otimes nu, alpha \otimes beta).
+def approx_kl(p, q, nx=None):
+    if nx is None:
+        p, q = list_to_array(p, q)
+        nx = get_backend(p, q)
+
+    return nx.sum(p * nx.log(p + 1.0 * (p == 0))) - nx.sum(p * nx.log(q))
+
+
+def product_kl(mu, nu, alpha, beta, nx=None):
+    """
+    Calculate the KL divergence between two product measures:
+    KL(mu \otimes nu, alpha \otimes beta) =
+    m_mu * KL(nu, beta) + m_nu * KL(mu, alpha) +
+    (m_mu - m_alpha) * (m_nu - m_beta)
+
+    Parameters
+    ----------
+    mu: vector or matrix
+    nu: vector or matrix
+    alpha: vector or matrix with the same size as mu
+    beta: vector or matrix with the same size as nu
+
+    Returns
+    ----------
+    KL divergence between two product measures
+    """
+
+    def kl(p, q, nx=None):
+        return approx_kl(p, q, nx=nx) - nx.sum(p) + nx.sum(q)
+
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    m_mu, m_nu = nx.sum(mu), nx.sum(nu)
+    m_alpha, m_beta = nx.sum(alpha), nx.sum(beta)
+    const = (m_mu - m_alpha) * (m_nu - m_beta)
+    res = m_nu * kl(mu, alpha, nx) + m_mu * kl(nu, beta, nx) + const
+
+    return res
+
+
+def product_l2(mu, nu, alpha, beta, nx=None):
+    """
+    norm = ||mu \otimes nu - alpha \otimes beta ||^2
+    = ||a||^2 ||b||^2 + ||mu||^2 ||nu||^2 - 2 < alpha, mu > < beta, nu >.
+    L2(mu \otimes nu, alpha \otimes beta) = norm / 2.
+    """
+
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    norm = nx.sum(alpha**2) * nx.sum(beta**2) \
+        - 2 * nx.sum(alpha * mu) * nx.sum(beta * nu) \
+        + nx.sum(mu**2) * nx.sum(nu**2)
+
+    return norm / 2
+
+
+def product_div(mu, nu, alpha, beta, divergence, nx=None):
+    if nx is None:
+        mu, nu, alpha, beta = list_to_array(mu, nu, alpha, beta)
+        nx = get_backend(mu, nu, alpha, beta)
+
+    if divergence == "kl":
+        return product_kl(mu, nu, alpha, beta, nx)
+    elif divergence == "l2":
+        return product_l2(mu, nu, alpha, beta, nx)
+
+
+# Support functions for BCD schemes
+def local_cost(data, pi, tuple_p, hyperparams, divergence, reg_type, nx=None):
```

Comment:
doc to do + change name so that we understand that it is related to uot. If it is simply a linearization of the cost, should be close to `gwgrad`